File: Proj4/Proj4a/e3/QuickSort.java
public class QuickSort {
    int partition(int arr[], int baixo, int alto) {
        int pivot = arr[alto];
        int i = (baixo-1);
        for (int j=baixo; j<alto; j++) {
            if (arr[j] <= pivot) {
                i++;
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        int temp = arr[i+1];
        arr[i+1] = arr[alto];
        arr[alto] = temp;
        return i+1;
    }

    void sort(int arr[], int baixo, int alto) {
        if (baixo < alto) {
            int pi = partition(arr, baixo, alto);
            sort(arr, baixo, pi-1);
            sort(arr, pi+1, alto);
        }
    }

    public static void main(String args[]) {
        Scanner scanner = new Scanner(System.in);
        int arr[10000];
        for (int i = 0; i < 10000; i++) {
              arr[i] = scanner.nextInt();
              if (entrada.isEmpty()) {
                  break;
              }
        }
        int n = arr.length;
        QuickSort ob = new QuickSort();
        ob.sort(arr, 0, n-1);
        for (int i=0; i<n; ++i)
            System.out.print(arr[i] + " ");
        System.out.println();
    }
}

File: Proj4/Proj4a/e3/quicksort.c
#include <stdio.h>

void swap(int* a, int* b) {
    int t = *a;
    *a = *b;
    *b = t;
}

int partition(int arr[], int baixo, int alto) {
    int pivot = arr[alto];
    int i = (baixo - 1);

    for (int j = baixo; j <= alto - 1; j++) {
        if (arr[j] < pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[alto]);
    return (i + 1);
}

void quick_sort(int arr[], int baixo, int alto) {
    if (baixo < alto) {
        int pi = partition(arr, baixo, alto);
        quickSort(arr, baixo, pi - 1);
        quickSort(arr, pi + 1, alto);
    }
}

int main() {
    int arr[10000];
    fgets(arr, sizeof(arr), stdin);
    for (int i = 0;i< 10000;i++)
    {
        if (arr[i] == '\n') arr[i] = '\0';
    }
    int n = sizeof(arr) / sizeof(arr[0]);
    quick_sort(arr, 0, n - 1);
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
    return 0;
}
